filedistribution/src/main/java/com/yahoo/vespa/filedistribution/maintenance/FileDistributionCleanup.java

```diff
@@ -0,0 +1,106 @@
+// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+package com.yahoo.vespa.filedistribution.maintenance;
+
+import com.yahoo.io.IOUtils;
+import java.io.File;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.time.Clock;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.stream.Stream;
+
+import static java.nio.file.Files.readAttributes;
+
+/**
+ * Removes file references not used since a configured time, but always keeps a certain number of file references
+ * even when they are unused (unused is based on last access time for the file).
+ *
+ * @author hmusum
+ */
+public class FileDistributionCleanup {
+
+    private static final Logger log = Logger.getLogger(FileDistributionCleanup.class.getName());
+    private static final int numberToAlwaysKeep = 20;
+
+    private final Clock clock;
+
+    public FileDistributionCleanup(Clock clock) {
+        this.clock = clock;
+    }
+
+    public List<String> deleteUnusedFileReferences(File fileReferencesPath,
+                                                   Duration keepFileReferencesDuration,
+                                                   Set<String> fileReferencesInUse) {
+        return deleteUnusedFileReferences(fileReferencesPath,
+                                          keepFileReferencesDuration,
+                                          numberToAlwaysKeep,
+                                          fileReferencesInUse);
+    }
+
+    public List<String> deleteUnusedFileReferences(File fileReferencesPath,
+                                                   Duration keepFileReferencesDuration,
+                                                   int numberToAlwaysKeep,
+                                                   Set<String> fileReferencesInUse) {
+        log.log(Level.FINE, () -> "Keep unused file references for " + keepFileReferencesDuration);
+        if (!fileReferencesPath.isDirectory()) throw new RuntimeException(fileReferencesPath + " is not a directory");
+
+        log.log(Level.FINE, () -> "File references in use : " + fileReferencesInUse);
+
+        Stream<String> candidates = sortedUnusedFileReferences(fileReferencesPath.toPath(), fileReferencesInUse, keepFileReferencesDuration);
```

Comment:
Add a try-with-resources to ensure the directory file descriptor is finally closed.

Reply:
Thanks, fixed